Add soil temperature and moisture IAU

tsga opened the pull request:
Add Incremental Analysis Update (IAU) to soil temperature and moisture in the NoahMP land model of the UWM's CCPP. Addresses issue: ufs-community/ccpp-physics/issues/190

Regression tests for ATM only and coupled (S2S) configurations pass in hera.

Additional land-IAU specific test to be added: control_c48_intel_land_iau (based on the "control_c48_intel" test, with the only difference being the land IAU increment files added to the INPUT directory . 

The input.nml is modified to add "land_iau_nml" (do_land_iau=.true. if executing with land IAU)

tsga committed:
land iau init

tsga committed:
mv iau forcing read to noahmpdrv_timestep_init

tsga committed:
Merge branch 'ufs-community:ufs/dev' into feature/lnd_iau

tsga committed:
add land iau as noahmpdrv_time_vary module

tsga committed:
Merge branch 'feature/lnd_iau' of https://github.com/tsga/ccpp-physics into feature/lnd_iau

* 'feature/lnd_iau' of https://github.com/tsga/ccpp-physics:
  correct the meta file
  correctly convert from flashes per five minutes to flashes per minute
  In physics/Interstitials/UFS_SCM_NEPTUNE/maximum_hourly_diagnostics.F90, scale lightning threat from flashes per 5 minutes to flashes per minute to match units in metadata
  In physics/Interstitials/UFS_SCM_NEPTUNE/maximum_hourly_diagnostics.meta: change units 'flashes 5 min-1' to 'flashes min-1' and update long name to make clear this is per 5 minutes

Commit:
revert to using noahmpdrv_timestep_init

Commit:
read all increment files at _init time

tsga committed:
debug

tsga committed:
Merge branch 'ufs-community:ufs/dev' into feature/lnd_iau

Commit:
deallocate at noahmpdrv_finalize

Commit:
comment nc mod

tsga committed:
Merge branch 'ufs-community:ufs/dev' into feature/lnd_iau

tsga committed:
mv config up

tsga committed:
delte _time_vary mods

Commit:
add nc90 funcs, cleanup, add comments

Commit:
add soil temp adjustments

Commit:
read fv3 increments

Commit:
update driver_timestepinit

Commit:
remove duplicte names

Commit:
remove duplicte names

Commit:
remove duplicte names

tsga committed:
fix arg_table_noahmpdrv_finalize

tsga committed:
Merge branch 'tmp' into feature/lnd_iau

* tmp:
  fix arg_table_noahmpdrv_finalize

Commit:
debug

Commit:
fix error about horizontal dimention

Commit:
fix error about horizontal dimention

Commit:
fix netcdf error

Commit:
fix netcdf error

Commit:
fix smc adjustment error

Commit:
fix smc adjustment error

Commit:
fix namelist typo

Commit:
fix netcdf dim error

Commit:
fix netcdf dim error

Commit:
remove Gaussian files

Commit:
remove Gaussian files

Commit:
remove Gaussian files

Commit:
debug memory

Commit:
fix fv3 file error

Commit:
fix fv3 file error

Commit:
fix fv3 file error

Commit:
add land iau conditions

Commit:
add land iau conditions

Commit:
add land iau conditions

Commit:
temp comment soilt consistency

Commit:
temp comment soilt consistency

Commit:
temp comment soilt consistency

Commit:
temp comment soilt consistency

Commit:
temp comment soilt consistency

Commit:
fix rdt error

Commit:
-

Commit:
-

Commit:
-

Commit:
-

Commit:
-

Commit:
debug

Commit:
clean up

Commit:
read mask from file

Commit:
revert back to calculating mask

Commit:
delete sim_nc

Commit:
use explcit array length

Commit:
use explcit array length

Commit:
debug print

Commit:
read land snow mask from inc files

Commit:
read land snow mask from inc files

Commit:
read land snow mask from inc files

Commit:
test on adj

Commit:
print debug info

Commit:
print debug info

Commit:
print debug info

Commit:
restrore stc/slc adjustments

Commit:
print diff indices

Commit:
bypass _timestep_init

Commit:
bypass _timestep_init

Commit:
test with hardcoded inc

Commit:
test with hardcoded inc

Commit:
test with hardcoded inc

Commit:
test with hardcoded inc

Commit:
test with hardcoded inc

Commit:
test non-iau increment hardcoded

Commit:
test 0 inc double prec

Commit:
remove hard-coded test

Commit:
remove debug prints

Commit:
add stc update and adjustment

Commit:
add stc update and adjustment

Commit:
zero out too small increments

Commit:
zero out too small increments

Commit:
add comment for single increment

Commit:
include t2 in update iau call

Commit:
set hr6 the only incr file (for testing)

Commit:
add increments at timestep_finalize (for testing)

Commit:
add sec argtable

Commit:
fix argtable

Commit:
input line in namelist for stcsmc adjustment

Commit:
clean up

Commit:
fix missing error code initialization

tsga committed:
Merge branch 'ufs-community:ufs/dev' into feature/lnd_iau

tsga committed:
remove namelist filename from iau struct

tsga committed:
use defaults when lnd_iau_nml doesn't exist

tsga commented:
Can't assign so tagging relevant people here. Please add (tag) anyone you think should be aware.
@CoryMartin-NOAA, @barlage, @ClaraDraper-NOAA, @CatherineThomas-NOAA, @junwang-noaa

grantfirl commented:
@tsga 2 of the 5 folks that you've requested to be added as reviewers have been added. The other 3 (@CoryMartin-NOAA, @ClaraDraper-NOAA, @CatherineThomas-NOAA) need to accept an invitation that I just sent to be collaborators on this repo before they can be added as reviewers.

Is there specific urgency for this work to be added to the UFS merge queue? We're trying to prioritize code reviews for a couple of "heavy" PRs.

ClaraDraper-NOAA commented:
@grantfirl Tseganeh is on annual leave, so I'll jump in. This PR is needed for testing of the land DA for GFSv17, which we hope to start in a few weeks (but could use a branch initially, if we must). It will need to be merged before we test the land DA in the formal DA prototypes (not possible until we have ensembles in the prototype).

grantfirl commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -9,99 +9,469 @@
+
+end subroutine noahmpdrv_timestep_init
+
+  !> \ingroup NoahMP_LSM
```
This CCPP phase (timestep_finalize) subroutine can be removed since it is effectively empty.

grantfirl commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -9,99 +9,469 @@
 
-      public :: noahmpdrv_init, noahmpdrv_run
+    public :: noahmpdrv_init, noahmpdrv_run, &
+              noahmpdrv_timestep_init, noahmpdrv_timestep_finalize, noahmpdrv_finalize
```
The timestep_finalize subroutinee can be removed since it is empty. One doesn't need to have subroutines present for CCPP phases that are not used in a scheme.

grantfirl commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -9,99 +9,469 @@
+    ! time step informatoin, etc)    
+    type (land_iau_control_type)          :: Land_IAU_Control
+    ! Land IAU Data holds spatially and temporally interpolated soil temperature increments per time step
+    type (land_iau_external_data_type)    :: Land_IAU_Data   !(number of blocks):each proc holds nblks
```
@dustinswales Could you help me to understand whether this will cause an issue similar to https://github.com/ufs-community/ccpp-physics/issues/74?

dustinswales replied:
@grantfirl I think it will. The solution for the photochemical schemes was to bundle everything up into a type and pass it in, so that each instance had a copy. But those are tiny schemes.

Another solution is to move its declaration to the module where it's defined, and set it up during initialization. For example, see how the [RRTMGP longwave driver](https://github.com/NCAR/ccpp-physics/blob/main/physics/Radiation/RRTMGP/rrtmgp_lw_main.F90#L16) sets up the "scheme native" [gas-optics DDT](https://github.com/NCAR/ccpp-physics/blob/main/physics/Radiation/RRTMGP/rrtmgp_lw_gas_optics.F90#L21). This approach should work as well, but keep in mind that the GP initialization routines are embedded with MPI commands, which aren't present in these type initializations. Also, this assumes that the data is the same across all instances, which it is for GP.

Unfortunately we don't have a way to test this, so we should at some point brainstorm a way to build a way to test this.

dustinswales replied:
^This is straightforward, but requires some effort.

grantfirl replied:
@dustinswales In your opinion, should this be done before this PR is approved/merged? I'm guessing that NRL is using NoahMP and therefore merging these changes would break their workflow? If so, we will probably need to provide some more guidance to move this along. I think that we need to come up with an implementation plan that we'd like to see followed. Also, I can't recall if we added something in the CCPP documentation about this issue, but we probably should so that we can refer to it and to help developers avoid this problem.

climbfuji replied:
We use NoahMP for testing. Can you help me refresh my memory what the problem is and if there is something I can do w.r.t. to testing?

dustinswales replied:
@grantfirl Tough call, since this is an NRL specific need. 
Does NEPTUNE use NoahMP? If not, then we don't need to do anything. If so, then we should discuss a plan with them (@climbfuji)?

grantfirl replied:
The issue is related to multiple instances of CCPP in one executable and the use of heap memory: https://github.com/ufs-community/ccpp-physics/issues/74

climbfuji replied:
See above. We have suites that use NoahMP for testing; the first operational implementation will use Noah.

If we can avoid breaking NoahMP for NEPTUNE, that would be great. What exactly is done here? Seeing this discussion/PR for the first time.

climbfuji replied:
Ok, I see it now. Those would have to be dimensioned as number of ccpp instances, or come from the host model

dustinswales replied:
@climbfuji The problem is related to data being stored on the HEAP when NEPTUNE has multiple ccpp instances in the same executable. A solution was to move these pieces from module memory and add them to the hosts typedefs, which I assume works for NEPTUNE? (We can't test this)

NoahMP has instances of its scheme-native DDTs stored on the HEAP. Will this cause a similar problem in NEPTUNE w/ multiple CCPP instances? I brought up RRTMGP because in there we use scheme-native DDTs, which are setup during initialization, so we are essentially doing the same thing as NoahMP, which may not work.
When NEPTUNE uses multiple instances, does each instance call ccpp init?

climbfuji replied:
Yes, each instance calls ccpp_init

dustinswales replied:
That's what I thought.
So if the writing to the DDTs was guarded by MPI directives, like in RRTMGP, we should be fine. No? (there is only one master processor...)

climbfuji replied:
For my understanding. Are these DDTs instance-dependent, i.e. are they potentially different for each instance, or are they identical regardless of the instance?

dustinswales replied:
In the case of the radiation, identical.
I can't speak for the LSM.

climbfuji replied:
This determines the approach we need to take. If the DDT is identical across all instances (e.g. containing static or time-dependent lookup tables etc.), then there can be one copy on the HEAP and initialization/destruction is done by the first instance but not the others.

If the DDT varies across instances, it either needs to be dimensioned by the number of CCPP instances, or it needs to be stored in the host model and passed as a variable.

tsga replied:
> In the case of the radiation, identical.
> I can't speak for the LSM.

@dustinswales different instances of the NoahMP LSM are likely to have different DDT values.

grantfirl replied:
Thanks @tsga, @climbfuji, @dustinswales for the discussion. I think that I understand the solution well enough to implement this here and in https://github.com/NOAA-EMC/fv3atm/pull/866. @tsga Would you be comfortable with me implementing this change and submitting PRs into [tsga:feature/lnd_iau](https://github.com/tsga/ccpp-physics/tree/feature/lnd_iau) and the associated fv3atm branch? Or, would you like to give it a shot?

grantfirl replied:
@tsga I guess that I should ask first, is there any reason why the instances of the types `land_iau_control_type, land_iau_external_data_type` can't be allocated in, e.g. GFS_typedefs.F90, and passed in? It's still fine to initialize them in the physics, but the memory just needs to "reside" in the host.

tsga replied:
@grantfirl, I can't see any reason against allocating them in the GFS_typedefs.F90.

tsga replied:
> Thanks @tsga, @climbfuji, @dustinswales for the discussion. I think that I understand the solution well enough to implement this here and in [NOAA-EMC/fv3atm#866](https://github.com/NOAA-EMC/fv3atm/pull/866). @tsga Would you be comfortable with me implementing this change and submitting PRs into [tsga:feature/lnd_iau](https://github.com/tsga/ccpp-physics/tree/feature/lnd_iau) and the associated fv3atm branch? Or, would you like to give it a shot?

@grantfirl, yes I would be happy with you implementing it. Thank you!

dustinswales replied:
@grantfirl Can you check if there is any performance impact when you are doing this? Just curious...

grantfirl commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,879 @@
+end subroutine land_iau_mod_set_control
+
+subroutine land_iau_mod_init (Land_IAU_Control, Land_IAU_Data, errmsg, errflg)     !nlunit, ncols, IPD_Data,,Init_parm)
+   ! integer,                              intent(in) :: me, mpi_root
```
Is there any good reason to keep the commented out variable declarations?

tsga replied:
> Is there any good reason to keep the commented out variable declarations?

not anymore.

grantfirl commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,879 @@
+   integer                                    :: input_nml_file_length    !< length(number of lines) in namelist for internal reads
+   character(len=4)                           :: iosstr
+
+   !> these are not available through the CCPP interface so need to read them from namelist file
```
If it is preferrable that  these variables are included in the existing `gfs_physics_nml` and passed in through the CCPP (like other iau namelist variables), then we can do that too. There's nothing wrong with reading in nml from a CCPP scheme in the init phase, but if we want to be consistent with past iau code, we can change it. Let me know if you would rather do that and whether you need help with it.

tsga replied:
@grantfirl I now see that separate nml for land iau would be better, so no need to include in gfs_physics nml. I will get rid of the old comment.

dustinswales commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -9,99 +9,469 @@
+    ! land iau mod
+    integer,                       intent(in) :: mpi_root   ! = GFS_Control%master        
+    character(*),                  intent(in) :: fn_nml
+    character(len=:), intent(in), dimension(:), pointer :: input_nml_file 
```
Remove pointer attriburte.

tsga committed:
Update noahmpdrv.F90

remove empty subroutine noahmpdrv_timestep_finalize

tsga committed:
Update noahmpdrv.meta

remove empty subroutine: noahmpdrv_timestep_finalize

tsga committed:
Update noahmpdrv.F90

removed pointer attribute from "input_nml_file" declaration

tsga committed:
Update lnd_iau_mod.F90

tsga committed:
Update lnd_iau_mod.F90

remove commented out old lines

tsga committed:
Update noahmpdrv.F90

tsga committed:
Update lnd_iau_mod.F90

get rid unused var input_nml_length

rhaesung commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -9,99 +9,446 @@
+  integer, allocatable          :: mask_tile(:)
+  integer,allocatable           :: stc_updated(:), slc_updated(:)
+  logical                       :: soil_freeze, soil_ice
+  integer                       :: n_freeze, n_thaw
```
Are these going to be used?

tsga replied:
They would be used if we are updating soil moisture (for now we are adding only temperature increments). I am commenting them out for now, but if they don't get used soon will delete them.

rhaesung commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -9,99 +9,446 @@
+  ! delt=GFS_Control%dtf
+  if ((Land_IAU_Control%dtp - delt) > 0.0001) then 
+    if(Land_IAU_Control%me == Land_IAU_Control%mpi_root) then 
+      print*, "Warning noahmpdrv_run delt ",delt," different from Land_IAU_Control%dtp ",Land_IAU_Control%dtp
```
In the warning message, should noahmpdrv_run be noahmpdrv_timestep_init?

tsga replied:
Yes, fixed.

rhaesung commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.meta:
```diff
@@ -96,8 +97,270 @@
+  intent = in
+[input_nml_file]
+  standard_name = filename_of_internal_namelist
+  long_name = amelist filename for internal file reads 
```
amelist -> namelist

tsga replied:
fixed, thank you.

rhaesung commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.meta:
```diff
@@ -96,8 +97,270 @@
+  intent = in
+[nx]
+  standard_name = number_of_points_in_x_direction_for_this_mpi_rank
+  long_name = number of points in x direction for this MPI rank
```
For consistency in naming, please consider changing it to "number of points in the x direction".

tsga replied:
done

rhaesung commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.meta:
```diff
@@ -96,8 +97,270 @@
+  intent = in
+[ny]
+  standard_name = number_of_points_in_y_direction_for_this_mpi_rank
+  long_name = number of points in y direction for this MPI rank
```
same here

rhaesung commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.meta:
```diff
@@ -96,8 +97,270 @@
+  intent = in
+[km]
+  standard_name = vertical_dimension_of_soil
+  long_name = soil vertical layer dimension
```
Please consider rephrasing to "vertical dimension of soil layers" for consistency with naming conventions used elsewhere.

tsga replied:
done

rhaesung commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.meta:
```diff
@@ -96,8 +97,270 @@
+  intent= in
+[weasd]
+  standard_name = water_equivalent_accumulated_snow_depth_over_land
+  long_name = water equiv of acc snow depth over land
```
Please consider changing it to "water equivalent of accumulated snow depth over land" for clarity and consistency.

rhaesung commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,848 @@
+               start = (/is, js, ks/), count = (/ix, jy, kz/))
+               ! start = (/is, js, ks/), count = (/ie - is + 1, je - js + 1, ke - ks + 1/))
+
+   end subroutine get_var3d_values
```
Missed error handling?

tsga replied:
I am passing on the only variable holding error info here (status) onto the calling function in lnd_iau module which then processes them.

rhaesung commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,848 @@
+
+   end subroutine get_var3d_values_int
+  
+end module land_iau_mod
```
Same here, missed the error handling?

tsga replied:
ditto, error handling done by netcdf_err

tsga committed:
Update noahmpdrv.F90

tsga committed:
Update noahmpdrv.F90

comment out unused number of freeze/thaw counters

tsga committed:
Update noahmpdrv.meta

fix typo

tsga committed:
Update noahmpdrv.meta

make long names consistent with naming conventions

climbfuji commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,848 @@
+
+  end type land_iau_control_type
+
+  type(land_iau_state_type) :: Land_IAU_state
```
@dustinswales @grantfirl Is this the problem?

grantfirl committed:
Merge branch 'ufs/dev' into feature/lnd_iau_gjf

grantfirl commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,848 @@
+
+  private
+
+  real(kind=kind_phys), allocatable :: wk3_stc(:, :, :, :), wk3_slc(:, :, :, :)
```
@tsga @dustinswales @climbfuji I've made modifications to move the module-defined DDTs to GFS_typedefs (in the FV3 host). I haven't tested/pushed the changes yet. The last bit that could potentially cause a problem is with these variables `wk3_stc, wk3_slc, wk3_slmsk`. They are read in during the init phase and used throughout the simulation in the timestep_init phase, so, basically like LUTs, but they are functions of time and x,y,z. I suppose that it's possible that different instances of the CCPP could read in different files and therefore these variables would be different across instances and therefore would need to get moved to the host as well. @tsga What do you think? Do you think that somewhere down the line (in potentially a different host model), that these variables would be different for each instance?

tsga replied:
@grantfirl yes, different instances are likely to have different wk3_stc and wk3_slc (but not wk3_slmsk).  Can these also be read by the host at init and passed through to noahmpdrv by the host? (The size of the time dim is often 3, so size of wk3_slc ~ 3*n_x*n_y*4. z=4 is for soil layers).  Alternatively, I can modify lnd_iau_mod.F90 to read the values from file (during timestep_init) then interpolate and store intermediate values in "Land_IAU_state". BTW, is it correct to assume that Land_IAU_state is not affected by the memory issue?

grantfirl replied:
See https://github.com/tsga/ccpp-physics/pull/1. Any module-level variables that need to persist when descoped need to be declared by the host unless those variables can effectively be shared across instances. Since the `wk3*` variables just contain the data read in by the host in the sfc_inc files (correct?), I'm not sure that they would need to be declared by the host. I.e., if different instances read in different sfc_inc files, then, yes, but otherwise, I think no.

grantfirl committed:
move declaration of land_iau_mod DDT instances from CCPP physics to host model

grantfirl committed:
add metadata for land IAU types

grantfirl committed:
fix compilation errors

grantfirl commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,848 @@
+   logical                                    :: exists
+   character(len=512)                         :: ioerrmsg
+
+   character(len=:), dimension(:)             :: input_nml_file => null()
```
I think that this needs to have the pointer attribute too in order to compile.

grantfirl replied:
See https://github.com/tsga/ccpp-physics/pull/1

grantfirl commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -9,99 +9,446 @@
+    ! land iau mod
+    integer,                       intent(in) :: mpi_root   ! = GFS_Control%master        
+    character(*),                  intent(in) :: fn_nml
+    character(len=:), intent(in), dimension(:) :: input_nml_file 
```
This needs to be declared with the pointer attribute in order to compile. See https://github.com/tsga/ccpp-physics/pull/1

grantfirl commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,848 @@
+module land_iau_mod
+  
+  use machine,                  only: kind_phys, kind_dyn
+  use physcons,                 only: pi => con_pi
```
I removed this in https://github.com/tsga/ccpp-physics/pull/1 since `pi` isn't used and this isn't how physical constants are supposed to be passed in CCPP schemes.

grantfirl commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,848 @@
+
+   ! Read all increment files at iau init time (at beginning of cycle) 
+   ! allocate (wk3_stc(n_t, 1:im,jbeg:jend, 1:km))  
+   call read_iau_forcing_fv3(Land_IAU_Control, errmsg, errflg)  !, wk3_stc, wk3_slc
```
I didn't address this in https://github.com/tsga/ccpp-physics/pull/1, but the code isn't checking for errors after returning from `read_iau_forcing_fv3`. When I try to run the new test with the data missing, instead of exiting with a nice error message, you get a seg fault because this code just continues and tries to access variables that were never allocated when the data files are missing.

grantfirl commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -9,99 +9,446 @@
-          return
-        end if
+    ! Read Land IAU settings 
+    call land_iau_mod_set_control(Land_IAU_Control, fn_nml, input_nml_file, &
```
IMO, it's a little awkward that the control for this functionality isn't read in until the physics init time. This means that every physics suite that uses NoahMP will at least have to go into the `land_iau_mod_set_control` routine to read the namelist. It would probably be cleaner to at least read in the do_land_iau variable in the host so that the physics can avoid doing anything with land IAU when it's not being used. I'm not aware of other physics schemes that force the host to read a namelist within the physics in order to know whether an option is used or not. I know that there is a GWD scheme that has a separate nml, but I think that there is also a namelist variable in the "main" physics namelist that controls whether the scheme namelist is even read. @dustinswales Do you have an opinion here?

tsga replied:
@grantfirl I have now moved the "land_iau_mod_set_control" call to "sfcprop_create". The land IAU namelist is also read here. I considered reading the land IAU specs in GFS_Control (similarly to that of the atm), but the sfcprop_create occurs in close proximity to GFS_Control_init, so I thought it is better to keep all land IAU related namelist reading and setup in one place.

grantfirl commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -9,99 +9,446 @@
+  
+  ! real (kind=kind_phys), dimension(max_soiltyp)  :: maxsmc, bb, satpsi
+  ! real, dimension(30)      :: maxsmc, bb, satpsi
+  real(kind=kind_phys), parameter          :: tfreez=273.16 !< con_t0c  in physcons
```
Why are these physical constants being redefined here?

tsga committed:
Merge pull request #1 from grantfirl/feature/lnd_iau_gjf

Move DDTs to host and fix compilation errors

Commit:
set land_iau_control from host

Commit:
calculate snowsoil mask at runtime

Commit:
combine DDTs holding increments; get rid of scheme level global array

Commit:
modify subroutines set/update increments

Commit:
default weight factors

Commit:
move weight factors to _IAU_Data

Commit:
update noahmpdrv meta

Commit:
fix time interval bounds

Commit:
handle valid time range better

Commit:
minor edit

Commit:
do netcdf error handling inside get_var3d_values

Commit:
error handling for read_iau_forcing_fv3

Commit:
remove redeclared constants in _timestep_int

Commit:
fix compilation errors

CoryMartin-NOAA commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,915 @@
+   logical               :: land_iau_do_stcsmc_adjustment = .false.
+   real(kind=kind_phys)  :: land_iau_min_T_increment = 0.0001
+  
+   NAMELIST /land_iau_nml/ do_land_iau, land_iau_delthrs, land_iau_inc_files, land_iau_fhrs,   &  !land_iau_gaussian_inc_file,   &
```
Remove the `gaussian_inc_file` comment?

CoryMartin-NOAA commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,915 @@
+   ! if (file_exist(fn_nml)) then 
+   inquire (file=trim(fn_nml), exist=exists)    ! TBCL: this maybe be replaced by nlunit passed from ccpp
+   if (.not. exists) then
+      write(6,*) 'lnd_iau_mod_set_control: namelist file ',trim(fn_nml),' does not exist'      
```
could you not define `errmsg` first and then `write(6,*) errmsg` since they are duplicated?

CoryMartin-NOAA commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,915 @@
+         if (me == mpi_root) then
+           write(6,*) 'lnd_iau_mod_set_control: I/O error code '//trim(iosstr)//' at land_iau namelist read' 
+         endif         
+         errmsg = 'lnd_iau_mod_set_control: I/O error code '//trim(iosstr)//' at land_iau namelist read'  
```
similar comment about above:
```
if (me == mpi_root) then
  write(6,*) errmsg
endif
```

CoryMartin-NOAA commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,915 @@
+   !nblks = Land_IAU_Control%nblks
+   !blksz = Land_IAU_Control%blksz(1)
+
+   print*, "rank is ie js je nlon nlat", Land_IAU_Control%me, is, ie, js, je, nlon, nlat
```
is this a debug statement? if not, consider making it a bit more verbose to describe what routine this is coming from?

CoryMartin-NOAA commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,915 @@
+            wt = 1.0
+         endif
+         normfact = normfact + wt
+         if (Land_IAU_Control%me == Land_IAU_Control%mpi_root) print *,'filter wts',k,kstep,wt
```
Similar comment, may want to specify this is from the land IAU module

CoryMartin-NOAA commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,915 @@
+
+   ! increment files in fv3 tiles 
+   if (trim(Land_IAU_Control%iau_inc_files(1)) .eq. '' .or. Land_IAU_Control%iaufhrs(1) .lt. 0) then ! only 1 file expected
+      print*, "Error! in Land IAU init: increment file name is empty or iaufhrs(1) is negative"
```
similar comment as previously here, also for consistency, probably should use either `print *,` or `write(6,*)` when writing similar lines, not both

CoryMartin-NOAA commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,915 @@
+      print*, "Error! in Land IAU init: increment file name is empty or iaufhrs(1) is negative"
+      errmsg = "Error! in Land IAU init: increment file name is empty or iaufhrs(1) is negative"
+      errflg = 1
+      ! Land_IAU_Control%do_land_iau=.false.
```
remove this commented out line?

CoryMartin-NOAA commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,915 @@
+   do k=1,ntimesall
+      if (Land_IAU_Control%iaufhrs(k) .lt. 0) exit
+      if (Land_IAU_Control%me == Land_IAU_Control%mpi_root) then
+         print *,k, " fhour ", Land_IAU_Control%iaufhrs(k)
```
if not a debug statement, consider making this more verbose

CoryMartin-NOAA commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,915 @@
+      do k=1,ntimes-1
+         if (idt(k) .ne. Land_IAU_Control%iaufhrs(2)-Land_IAU_Control%iaufhrs(1)) then
+            print *,'in land_iau_init: forecast intervals in iaufhrs must be constant'
+         !   call mpp_error (FATAL,' forecast intervals in iaufhrs must be constant')
```
remove MPP commented out line?

CoryMartin-NOAA commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,915 @@
+   if (Land_IAU_Control%me == Land_IAU_Control%mpi_root) print *,'land_iau_init: ntimes = ',ntimes
+   Land_IAU_Control%ntimes = ntimes
+   if (ntimes < 1) then
+      print*, "Error! in Land IAU init: ntimes < 1"
+      errmsg = "Error! in Land IAU init: ntimes < 1"
+      errflg = 1
+      ! Land_IAU_Control%do_land_iau=.false.
```
Similar comments as before on the print statement + commented out line

CoryMartin-NOAA commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,915 @@
+   ! Read all increment files at iau init time (at beginning of cycle) 
+   ! allocate (wk3_stc(n_t, 1:im,jbeg:jend, 1:km))  
+   call read_iau_forcing_fv3(Land_IAU_Control, Land_IAU_state%stc_inc, Land_IAU_state%slc_inc, errmsg, errflg)  !, wk3_stc, wk3_slc
+   ! call read_iau_forcing_fv3(Land_IAU_Control, Land_IAU_state%inc1%stc_inc, Land_IAU_state%inc1%slc_inc, errmsg, errflg)
```
do we need this line?

CoryMartin-NOAA commented on physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90:
```diff
@@ -0,0 +1,915 @@
+   ! call read_iau_forcing_fv3(Land_IAU_Control, Land_IAU_state%inc1%stc_inc, Land_IAU_state%inc1%slc_inc, errmsg, errflg)
+   if (errflg .ne. 0) return
+   
+   ! increments already in the fv3 grid--no need for interpolation       
+   ! do k = 1, npz  ! do k = 1,n_soill    !  
+   !    do j = 1, nlat
+   !       do i = 1, nlon                
+   !          Land_IAU_state%inc1%stc_inc(i,j,k)  = wk3_stc(1, i, j, k)
+   !          Land_IAU_state%inc1%slc_inc(i,j,k)  = wk3_slc(1, i, j, k) 
+   !       end do
+   !     enddo
+   ! enddo
+
+   if (ntimes.EQ.1) then  ! only need to get incrments once since constant forcing over window
+      call setiauforcing(Land_IAU_Control, Land_IAU_Data, Land_IAU_state)
+      Land_IAU_Data%itnext = 0
+   endif
+   if (ntimes.GT.1) then  !have multiple files, but only need 2 at a time and interpoalte for timesteps between them  
+       ! interpolation is now done in land_iau_mod_getiauforcing 
+      Land_IAU_Data%hr2=Land_IAU_Control%iaufhrs(2)   
+      Land_IAU_Data%itnext = 2
+      ! ! Land_IAU_Data%snow_land_mask(:, :)  = wk3_slmsk(1, :, :)
+      ! do k = 1, npz  ! do k = 1,n_soill    !  
+      !    do j = 1, nlat
+      !       do i = 1, nlon                
+      !          Land_IAU_state%inc2%stc_inc(i,j,k)  = wk3_stc(2, i, j, k)
+      !          Land_IAU_state%inc2%slc_inc(i,j,k)  = wk3_slc(2, i, j, k) 
+      !       end do
+      !     enddo
+      ! enddo
+   endif
+!   print*,'end of IAU init',dt,rdt
```
remove all these interpolation related commented out lines?

CoryMartin-NOAA replied:
Or maybe just have a comment that says 'code assumes increments are on model native grid'

Commit:
clean up, remove debug print outs

barlage commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -690,11 +1050,6 @@ subroutine noahmpdrv_run                                       &
-!
-!  --- Just return if external land component is activated for two-way interaction
-!
-  if (cpllnd .and. cpllnd2atm) return
```
why are these lines removed?

tsga replied:
unless by mistake, I am not supposed to touch those. I will look.

tsga committed:
restore if (cpllnd .and. cpllnd2atm) check

Commit:
clean noahmpdrv

Commit:
Merge branch 'feature/lnd_iau' of https://github.com/tsga/ccpp-physics into feature/lnd_iau

Commit:
remove debug prints

tsga committed:
Update noahmpdrv.F90

tsga committed:
Update noahmpdrv.F90

tsga committed:
Update noahmpdrv.F90

tsga committed:
Update noahmpdrv.F90

tsga committed:
Update noahmpdrv.F90

tsga committed:
Update noahmpdrv.F90

tsga committed:
Update noahmpdrv.F90

tsga committed:
Update noahmpdrv.F90

tsga committed:
Update noahmpdrv.F90

Commit:
change DDTs in noahmpdrv_init to optional

Commit:
update noahmpdrv.meta for edits to accomodate component model NoahMP

Commit:
fix compile error

tsga committed:
Merge branch 'ufs-community:ufs/dev' into feature/lnd_iau

tsga committed:
fix real type for mask

grantfirl commented:
@tsga Are you finished making changes to this PR?

ClaraDraper-NOAA reviewed (approved):
Science-wise, looks good to me, thanks @tsga

tsga commented:
> @tsga Are you finished making changes to this PR?

@grantfirl yes. And the science reviewers are also OK with the changes. The RTs passed (with the -c switch for lnd_iau because its baseline is not on hera). Thank you.

tsga committed:
Merge branch 'ufs/dev' into feature/lnd_iau

* ufs/dev:
  fix another optional argument in RUC LSM
  fix some more optional variables
  fix some optional arugments

mkavulich commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -13,13 +13,20 @@ module noahmpdrv
       use module_sf_noahmplsm
 
+! Land IAU increments for soil temperature (plan to extend to soil moisture increments)
+      use land_iau_mod,  only: land_iau_control_type, land_iau_external_data_type,  &
```
Is there a reason this `use land_iau_mod` statement is split across two different commands?

tsga replied:
just a desire to separate data types from function/procedure. I have combined them now.

grantfirl commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.meta:
```diff
@@ -96,6 +96,230 @@
+  units = mixed
+  dimensions = ()
+  type = land_iau_control_type
+  intent = inout
```
Since these are declared with the optional attribute, you need to add an `optional = True` line below the intent line for each of these variables.

grantfirl commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -100,9 +119,309 @@ subroutine noahmpdrv_init(lsm, lsm_noahmp, me, isot, ivegsrc, &
+  ! GFS_control%jdat(:) = jdat(:)  
+  Land_IAU_Control%fhour=fhour
+  if(Land_IAU_Control%me == Land_IAU_Control%mpi_root) then 
+    print*,"itime ",itime," GFScont%fhour ",fhour," IauCon%fhour",Land_IAU_Control%fhour,   &
```
Please consider wrapping any `print` statements in a debug logical that one can set in code or elsewhere. There have been complaints from folks that physics schemes add too much to the output console, cluttering it up.

grantfirl commented on physics/SFC_Models/Land/Noahmp/noahmpdrv.F90:
```diff
@@ -100,9 +119,309 @@ subroutine noahmpdrv_init(lsm, lsm_noahmp, me, isot, ivegsrc, &
+    deallocate(mask_tile)
+  
+
+    write(*,'(a,i2)') ' noahmpdrv_timestep_init: statistics of grids with stc/smc updates for rank : ', Land_IAU_Control%me
```
Same thing. Please consider only outputting this when a debug flag is set.

grantfirl commented:
@tsga Sorry for the late re-review. Please see the comments above. The only thing that I think must be changed is the addition of the `optional = True` lines in the metadata. The other comments are to keep other physics developers happy. Also, I understand that this functionality has been under development so there are many lines of code commented here and there as needs changed. Given git history, it should be possible to recover those lines rather than leave them commented out in the code.

I feel comfortable approving once the noahmpdrv.meta has been changed.

tsga commented:
> @tsga Sorry for the late re-review. Please see the comments above. The only thing that I think must be changed is the addition of the `optional = True` lines in the metadata. The other comments are to keep other physics developers happy. Also, I understand that this functionality has been under development so there are many lines of code commented here and there as needs changed. Given git history, it should be possible to recover those lines rather than leave them commented out in the code.
> 
> I feel comfortable approving once the noahmpdrv.meta has been changed.

@grantfirl no problem, thank you for the feedback. I will address these soon and progress the PR forward.

Commit:
add optional=True for lndiau ddts

Commit:
clean lnd_iau_mod

Commit:
clean noahmpdrv_timestep_init

tsga commented:
@grantfirl I have now added the optional=True as you suggested, removed all commented lines that are not descriptive comments. I have also removed most of the print statements. Left few that I think are essential but they are now very few and most printed only once.

mkavulich reviewed (commented):
Just one question, otherwise looks good

Commit:
combine use lnd_iau_mod lines

FernandoAndrade-NOAA commented:
Testing for [2415](https://github.com/ufs-community/ufs-weather-model/pull/2415) has completed, please continue with the merge process, thank you.

grantfirl commented:
@tsga @mkavulich It looks like the last commit just got in under the deadline for final UWM tests, so I'm going to go ahead and merge this.

tsga commented:
> @tsga @mkavulich It looks like the last commit just got in under the deadline for final UWM tests, so I'm going to go ahead and merge this.

Thank you so much @grantfirl and everyone for guiding me through these PR. Cheers.
